map/src/main/kotlin/edu/artic/map/MapViewModel.kt

```diff
@@ -0,0 +1,200 @@
+                }
+
+        val allBuildingNames = mapAnnotationDao
+                .getBuildingNamesOnMap()
+                .map { annotationList ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    annotationList.forEach { annotation ->
+                        val floor = annotation.floor
+                                .let {
+                                    it?.toInt() ?: 1
+                                }
+                        list.add(MapItem.Annotation(annotation, floor))
+                    }
+                    return@map list
+
+                }
+
+        Observable.combineLatest(
+                allAmenities.asObservable(),
+                allBuildingNames.asObservable(),
+                BiFunction<List<MapItem.Annotation>, List<MapItem.Annotation>, List<MapItem.Annotation>>
+                { amenities, buildingNames ->
+                    val list = mutableListOf<MapItem.Annotation>()
+                    list.addAll(amenities)
+                    list.addAll(buildingNames)
+                    return@BiFunction list
+                }).bindTo(alwaysVisibleAnnotations)
+                .disposedBy(disposeBag)
+
+
+        setupZoomLevelOneBinds()
+        setupZoomLevelTwoBinds()
+        setupZoomLevelThreeBinds()
+
+    }
+
+
+    fun setupZoomLevelOneBinds() {
+        Observable.combineLatest(
+                zoomLevel.distinctUntilChanged().filter { it === MapZoomLevel.One },
+                floor.distinctUntilChanged(),
+                alwaysVisibleAnnotations,
+                Function3<MapZoomLevel, Int, List<MapItem.Annotation>, List<MapItem<*>>>
+                { _: MapZoomLevel, floor: Int, annotations: List<MapItem.Annotation> ->
+                    annotations.filter { it.item.floor?.toInt() == floor }
+                }).bindTo(mapAnnotations)
+                .disposedBy(disposeBag)
+    }
+
+    fun setupZoomLevelTwoBinds() {
+        Observable.combineLatest(
+                zoomLevel.distinctUntilChanged().filter { it === MapZoomLevel.Two },
+                floor.distinctUntilChanged(),
+                alwaysVisibleAnnotations,
+                mapAnnotationDao
+                        .getDepartmentOnMap()
+                        .asObservable()
+                        .map { departmentList ->
+                            val list = mutableListOf<MapItem.Annotation>()
+                            departmentList.forEach { department ->
+                                val floor = department.floor
+                                        .let {
+                                            it?.toInt() ?: 1
+                                        }
+                                list.add(MapItem.Annotation(department, floor))
+                            }
+                            return@map list
+
+                        },
+                Function4<MapZoomLevel, Int, List<MapItem.Annotation>, List<MapItem.Annotation>, List<MapItem<*>>>
```

Comment:
not sure we need explicit signatures here. are you not using RXKotlin? Observables.combineLatest()

Reply:
done